crates/qasm3/src/lib.rs

```diff
@@ -0,0 +1,207 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+mod build;
+mod circuit;
+mod error;
+mod expr;
+
+use std::ffi::OsString;
+use std::path::{Path, PathBuf};
+
+use hashbrown::HashMap;
+
+use pyo3::prelude::*;
+use pyo3::types::{PyModule, PyTuple};
+
+use oq3_semantics::syntax_to_semantics::parse_source_string;
+
+use crate::error::QASM3ImporterError;
+
+/// The name of a Python attribute to define on the given module where the default implementation
+/// of the ``stdgates.inc`` custom instructions is located.
+const STDGATES_INC_CUSTOM_GATES_ATTR: &str = "STDGATES_INC_GATES";
+
+/// Load an OpenQASM 3 program from a string into a :class:`.QuantumCircuit`.
+///
```

Comment:
This needs to have a warning about it being an experimental unstable interface.

Reply:
But it should say a bit more than that. It's not really experimental, it's just preliminary. Except for the name, we are committed to the interface.

Reply:
That's fair, "experimental api" is just the shorthand we've been using to describe this situation. The exact language I put in the stability policy isn't as specific as to what this situation is called, all it says is:

> For the purposes of semantic versioning, the Qiskit public API is considered any documented module, class, function, or method that is not marked as private (with an underscore _ prefix). However, there can be explicit exceptions made for specific documented APIs. In such cases, these APIs will be clearly documented as not being considered stable interfaces yet, and a user-visible warning will be actively emitted on any use of these unstable interfaces.

So we can call it whatever we want as long as we make the unstable nature of the interface explicit in the docstring. (we already have the warning)

Reply:
What I've written is like this:
```rust
/// .. warning::
///
///     This native version of the OpenQASM 3 importer is currently experimental.  It is typically
///     much faster than :func:`~qiskit.qasm3.loads`, but has a reduced supported feature set,
///     which will expand over time.
```
it's just not pushed because I'm also writing the release note.  Happy to change that if people prefer.

Reply:
It's faster and also gives better diagnostics. I imagine the latter might be more useful (But who knows). I don't know if this warning text is the place to say this.
But as long as you have "much faster", I think it's worth saying "much faster, and gives more informative error messages:.

Reply:
Done in 6e729bfdd7.  I hadn't included the bit about the error messages in line in the individual function docstrings, but it's in the module docs in fd80ea1b6c.